File: Easy/1863. Sum of All Subset XOR Totals.cpp
// Bit Manipulation
// Time compelxity: O(n)
// Space compelxity: O(1)
class Solution {
public:
    int subsetXORSum(vector<int>& nums) {
        int result = 0;

        // Capture each bit that is set in any of the elements
        for (int num : nums) {
            result = result | num;
        }

        // Multiply by the number of subsets XOR totals that will have each bit set
        return result << ((nums.size() - 1));
    }
};

File: Easy/1863. Sum of All Subset XOR Totals.java
// Optimized Backtraking
// Time Complexity: O(2^n)
// Space Complexity: O(n)
class Solution {
    public int subsetXORSum(int[] nums) {
        return XORSum(nums, 0, 0);
    }

    private int XORSum(int[] nums, int index, int currentXOR) {
        // Return currentXOR when all the elements are traversed
        if (index == nums.length)
            return currentXOR;

        // Calculate sum of subset XOR with current element
        int with_element = XORSum(nums, index + 1, currentXOR ^ nums[index]);

        // Calculate sum of subset XOR without current element
        int without_element = XORSum(nums, index + 1, currentXOR);

        // return sum of XOR totals
        return with_element + without_element;

    }
}
